relayer/src/services/relayer/arbitrum.ts

```diff
@@ -30,61 +36,101 @@ const L1_CONTRACTS: Record<number, { ARBITRUM_L1_MESSENGER: string }> = {
 export class ArbitrumRelayerService extends BaseRelayerService {
   static readonly MAX_MESSAGE_LENGTH_BYTES = 5_000;
 
-  async prepare(): Promise<boolean> {
-    console.log("Preparing Arbitrum Messenger for forwarding the L1 message...");
+  prepareStep = new PrepareStep(
+    async () => await this.prepare(),
+    async () => await this.isPrepareCompleted(),
+  );
+  relayTxToL1Step = new RelayStep(
+    async (l2TransactionHash) => await this.relayTxToL1(l2TransactionHash),
+    async (l2TransactionHash) => await this.isRelayCompleted(l2TransactionHash),
+  );
+  finalizeStep = new FinalizeStep(
+    async (l1TransactionHash) => await this.finalize(l1TransactionHash),
+    async (l1TransactionHash) => await this.isFinalizeCompleted(l1TransactionHash),
+  );
+
+  private arbitrumL1Messenger: ArbitrumL1Messenger;
+  private deltaSubmissionFee?: BigNumber;
+
+  private l1ToL2Message?: L1ToL2MessageWriter;
 
-    const arbitrumMessengerAddress = L1_CONTRACTS[await this.getL2ChainId()].ARBITRUM_L1_MESSENGER;
+  private l2ToL1Message?: L2ToL1MessageWriter;
 
-    const arbitrumL1Messenger = ArbitrumL1Messenger__factory.connect(
+  constructor(...args: ConstructorParameters<typeof BaseRelayerService>) {
+    super(...args);
+
+    const arbitrumMessengerAddress = L1_CONTRACTS[this.l2ChainId].ARBITRUM_L1_MESSENGER;
+    this.arbitrumL1Messenger = ArbitrumL1Messenger__factory.connect(
       arbitrumMessengerAddress,
       this.l1Wallet,
     );
+  }
 
-    const currentDeposit: BigNumber = await arbitrumL1Messenger.deposits(this.l1Wallet.address);
+  private async getDeltaSubmissionFee(): Promise<BigNumber> {
     const l1ToL2MessageGasEstimate = new L1ToL2MessageGasEstimator(this.l2RpcProvider);
-
     const submissionFee = await l1ToL2MessageGasEstimate.estimateSubmissionFee(
       this.l1RpcProvider,
       BigNumber.from(0),
       BigNumber.from(ArbitrumRelayerService.MAX_MESSAGE_LENGTH_BYTES),
     );
-
     const deltaSubmissionFee = submissionFee.mul(110).div(100); // We add 10% on top to ensure that the creation of a retryable ticket doesn't fail
 
-    if (currentDeposit.lte(deltaSubmissionFee)) {
-      // ArbitrumL1Messenger refunds the rest of the deposit automatically
-      // after relaying a message by calling `ArbitrumL1Messenger.sendMessage`.
-      const estimatedGasLimit = await arbitrumL1Messenger.estimateGas.deposit({
-        value: deltaSubmissionFee,
-      });
+    return deltaSubmissionFee;
+  }
 
-      const transaction = await arbitrumL1Messenger.deposit({
-        gasLimit: estimatedGasLimit,
-        value: deltaSubmissionFee,
-      });
+  private async isPrepareCompleted(): Promise<void | false> {
```

Comment:
The reason is that the `isCompleted` function of the relay step needs to return a value. I agree that a simple boolean would be much nicer. One option could be to separate the recovering of the hash into another function. Since this would be a bigger change, I will  try to fix the rest first and see if we can make this today.

Reply:
Okay, now i actually see what the problem was. Sorry that you had to go that far to fix such a small thing.
On one hand, the separation of the steps is much cleaner and easily understandable now, but we introduced 1 extra level of rpc calls but we shouldn't be too worried about it for now as this program is being ran rarely.